### ISSUE TITLE
[Bug] xgb_win_prob_model.json loaded from disk on every prediction click instead of once at startup — repeated disk I/O blocks prediction response

### ISSUE THREAD

**Issue by anshul23102** [Arnav-Singh-5080/IPL-Win-Probability-Prediction#100]
## Description

Inside the `if analyze:` prediction block in `application.py`, the XGBoost model is loaded from disk on **every single click of the Run Analysis button**:

```python
# application.py — inside `if analyze:` block
if analyze:
    runs_left = target - score
    balls_left = 120 - (overs * 6)
    # ...

    loaded_xgb = xgb.XGBClassifier()           # New object created every click
    loaded_xgb.load_model('xgb_win_prob_model.json')  # Disk read every click
```

Streamlit reruns the entire script on every user interaction. Because the model load is inside the `if analyze:` block rather than in a `@st.cache_resource`-decorated function, a fresh disk read and full model deserialization occurs every time the user clicks **Run Analysis**.

## Impact

1. **Prediction latency**: Each click requires reading and deserializing the XGBoost model JSON file before the prediction can run. On Streamlit Cloud (which uses shared I/O), this adds a measurable delay on every interaction — the model file is ~500KB-2MB and XGBoost deserialization is not instant.
2. **Scalability**: Under concurrent users, repeated model loads from disk compete for I/O bandwidth, degrading response times for all users simultaneously.
3. **Correctness conflict with `train_model()`**: The app already calls `@st.cache_resource` on `train_model()` to cache the sklearn pipeline, but uses no caching whatsoever for the model actually used in production predictions. This is inconsistent and confusing.

## Steps to Reproduce

1. Open the app and navigate to **Match Analysis**.
2. Click **Run Analysis** multiple times, changing inputs between clicks.
3. Each click triggers a fresh `load_model('xgb_win_prob_model.json')` call rather than reusing an in-memory instance.

## Proposed Fix

Move the model load to a `@st.cache_resource` function so it is executed once at startup and reused across all prediction calls:

```python
# At module level — runs once and is cached
@st.cache_resource
def load_xgb_model():
    model = xgb.XGBClassifier()
    model.load_model('xgb_win_prob_model.json')
    return model

xgb_model = load_xgb_model()
```

Then in the prediction block:

```python
if analyze:
    # Remove the two lines below:
    # loaded_xgb = xgb.XGBClassifier()
    # loaded_xgb.load_model('xgb_win_prob_model.json')

    # Use the pre-loaded cached model:
    proba = xgb_model.predict_proba(input_df)[0]
```

This ensures the model is deserialized exactly once per server process lifetime and all prediction calls reuse the same in-memory instance.

## Affected File

`application.py` — `if analyze:` block, lines ~1522-1524

## Suggested Labels

`type:bug` `level:intermediate` `quality:exceptional`

I would like to work on this issue under GSSoC 2026. Could you please assign it to me? @Arnav-Singh-5080

/assign

**Comment by anshul23102** [comment]
Opened PR #107 with the fix.

Root cause: `train_model()` was called at module level on every cold server start, reading two CSV files and training a LogisticRegression pipeline from scratch before any prediction could run. A pre-trained `pipe.pkl` and `joblib` (in requirements.txt) were present but unused.

Fix: added `load_pipeline()` decorated with `@st.cache_resource` that loads `pipe.pkl` via `joblib.load()`. The decorator ensures deserialization happens exactly once per server process and the in-memory pipeline is shared across all sessions and reruns with zero subsequent disk I/O. A `FileNotFoundError` fallback to `train_model()` keeps the app functional if the artifact is missing. The call site is updated from `pipe = train_model()` to `pipe = load_pipeline()`.

**Comment by SYEDABRAR037** [comment]
hi could you please assign  this issue to me so that i can  work on it
